### DISCUSSION TITLE
Updating kernel cmdline

### DISCUSSION THREAD

**Discussion by eraserix** [rauc/rauc#839] [Q&A]
I've recently started to experiment with RAUC. For tests I try to use a symmetric A/B setup with the following partitions:

1. grub for grubenv and grub.cfg
2. root A
3. root B
4. data

Now my question: How do I match kernel boot parameters to the deployed rootfs (=> the rootfs defines at least part of the kernel commandline). What I thought of so far:

- Compile them into the kernel (inflexible, needs kernel recompile whenever boot params change)
- Run some kind of hook that overwrites grub.cfg with new parameters on image deploy (I feel this is risky, but distros seem to do fine with that)
- Come up with some grub scripting that loads the parameters from the rootfs and add them to the "kernel" line

How do you guys do it?

Thanks!
Christoph

**Comment by ejoerns** [comment]
Normally, the bootloader specification config files are a good way to achieve what you intend.
https://systemd.io/BOOT_LOADER_SPECIFICATION/

However, I haven't heard this being supported for grub yet. Are you tied to grub? Or could you use another bootloader?

The other option I see is actually compiling this into the kernel, yes.

Updating a central grub.cfg from the individual installations sounds quite fragile to me.

Also consider that you might want to able to update your bootloader later in a fail-safe manner. For this, it would be required to separate grubenv from actual grub storage partition (and probably use the atomic GPT update feature of RAUC: https://rauc.readthedocs.io/en/latest/advanced.html#update-boot-partition-in-gpt)

**Reply by a3f** [reply]
> However, I haven't heard this being supported for grub yet

jfyi, Fedora patches GRUB to support bootloader spec.

**Reply by eraserix** [reply]
> Normally, the bootloader specification config files are a good way to achieve what you intend.
> https://systemd.io/BOOT_LOADER_SPECIFICATION/

From a quick read it looks like all systems manage their boot stuff (config, kernel, initrd) inside a single partition? This does not sound a whole lot different than fiddling about with a script that sources some parts of its configuration...

> However, I haven't heard this being supported for grub yet. Are you tied to grub? Or could you use another bootloader?

No, I'm not tied to grub. To be honest, I have not looked at the bootloader part at all, I just used what was already there. So I'm open to suggestions.

Our system is built from a customized Debian rootfs that is unpacked onto the target. We are on x86_64 and still use BIOS boot.
